### PR TITLE
add other collection measures

### PR DESCRIPTION
**Pull request by Martin-Hoppe**
list mapping
list weighted mapping
list correctness

### CODE REVIEW

File: cbrkit/sim/collections.py [hunk 1]
```diff
@@ -190,3 +190,202 @@ def wrapped_func(query: Sequence[ValueType], case: Sequence[ValueType]) -> float
...
+    if list_weights is None:
+        list_weights = []
+
+    list_mapping_func = list_mapping(
+        similarity_to_use, contains, return_local_similarities
+    )
+
+    def wrapped_func(
+        x: Collection[Any], y: Collection[Any]
+    ) -> tuple[float | Any, list[Any] | Any] | float | Any:
```
Comment (mirkolenz) [inline cbrkit/sim/collections.py:295]: The return type does not correspond to the `SimPairFunc` annotation.

**Review by mirkolenz — CHANGES_REQUESTED** [review]

File: cbrkit/sim/collections.py [hunk 2]
```diff
@@ -190,3 +190,202 @@ def wrapped_func(query: Sequence[ValueType], case: Sequence[ValueType]) -> float
...
+
+    return wrapped_func
+
+
+def list_mapping_weighted(
+    similarity_to_use: Callable[[Any, Any], float] = None,
+    list_weights: List[Dict[str, Any]] = None,
+    contains: str = "inexact",
+    return_local_similarities: bool = True,
+) -> SimPairFunc[Collection[Any], float]:
```
[the remark this answers is not in the document]
Comment (Martin-Hoppe) [reply cbrkit/sim/collections.py:270]: Having it as a standalone is closer to the original ProCake though, if thats irrelevant i can change it

[the remark this answers is not in the document]
Comment (mirkolenz) [reply cbrkit/sim/collections.py]: Yeah I would say that this is irrelevant, cbrkit is already quite different from ProCAKE, so please go ahead and make this a parameter for the `list_mapping` function 😄

File: cbrkit/sim/collections.py [hunk 3]
```diff
@@ -190,3 +192,227 @@ def wrapped_func(query: Sequence[ValueType], case: Sequence[ValueType]) -> float
         return best_score
 
     return wrapped_func
+
+
+@dataclass
+class ListMappingSim:
+    value: float
+    local_similarities: List[float] = None
```
Comment (mirkolenz) [inline cbrkit/sim/collections.py:200]: The types do not match, maybe you mean one of the following?

```suggestion
    local_similarities: List[float] | None = None
```
```suggestion
    local_similarities: List[float] = field(default_factory=list)
```

File: cbrkit/sim/collections.py [hunk 4]
```diff
@@ -190,3 +192,227 @@ def wrapped_func(query: Sequence[ValueType], case: Sequence[ValueType]) -> float
...
+
+
+def list_mapping(
+    element_similarity: Callable[[Any, Any], float] = None,
```
Comment (mirkolenz) [inline cbrkit/sim/collections.py:204]: Type does not correspond to default value.

File: cbrkit/sim/collections.py [hunk 5]
```diff
@@ -190,3 +192,227 @@ def wrapped_func(query: Sequence[ValueType], case: Sequence[ValueType]) -> float
...
+    exact: bool = False,
+    return_local_similarities: bool = False,
```
Comment (mirkolenz) [inline cbrkit/sim/collections.py:206]: Is this parameter still needed? Now that a dataclass is used (good!), the value could always be added. Or is there some performance impact?

File: cbrkit/sim/collections.py [hunk 6]
```diff
@@ -190,3 +192,227 @@ def wrapped_func(query: Sequence[ValueType], case: Sequence[ValueType]) -> float
...
+
+
+@dataclass
+class Weight:
+    weight: float
+    lower_bound: float
+    upper_bound: float
+    inclusive_lower: bool
+    inclusive_upper: bool
+    normalized_weight: float = None  # Will be set later
```
Comment (mirkolenz) [inline cbrkit/sim/collections.py:285]: Type does not correspond to default value.

```suggestion
    normalized_weight: float | None = None  # Will be set later
```

File: cbrkit/sim/collections.py [hunk 7]
```diff
@@ -190,3 +192,227 @@ def wrapped_func(query: Sequence[ValueType], case: Sequence[ValueType]) -> float
...
+            local_similarities=local_similarities
+            if return_local_similarities
+            else None,
+        )
+
+    return wrapped_func
+
+
+def list_correctness(
+    discordant_parameter: float = 1.0,
```
Comment (mirkolenz) [inline cbrkit/sim/collections.py:368]: The naming is a bit odd. What about `worst_case_sim` or `fallback_sim`.

Also, is a default value of `1.0` sensible here? Given the description, I would assume that `0.0` could also make sense.

**Review by mirkolenz — CHANGES_REQUESTED** [review]

**Review by mirkolenz — APPROVED** [review]

**Pull request merged**